python/tskit/combinatorics.py

```diff
@@ -272,15 +270,55 @@ def split_polytomies(
         if tree.num_children(u) > 2:
             root = TreeNode.random_binary_tree(tree.children(u), rng)
             root.label = u
-            time = tree.time(u) - epsilon
-            stack = [(child, time) for child in root.children]
+            root_time = tree.time(u)
+            stack = [(child, root_time) for child in root.children]
             while len(stack) > 0:
-                node, time = stack.pop()
+                node, parent_time = stack.pop()
                 if node.label is None:
-                    node.label = tables.nodes.add_row(time=time)
+                    if epsilon is None:
+                        child_time = np.nextafter(parent_time, -np.inf)
+                    else:
+                        child_time = parent_time - epsilon
+                    node.label = tables.nodes.add_row(time=child_time)
+                else:
+                    assert len(node.children) == 0
+                    # This is a leaf node connecting back into the original tree
+                    child_time = tree.time(node.label)
+                if parent_time <= child_time:
+                    u = root.label
+                    min_child_time = min(tree.time(v) for v in tree.children(u))
+                    min_time = root_time - min_child_time
+                    message = (
```

Comment:
Thanks for adding these messages in. This is really helpful. It would be even nicer for the user if we saved the "worst" message (ie. the example that would require the smallest epsilon to be set) until we have passed through all the nodes, and then output just that one after the loop (in the hope that correcting that one if possible would make the routine run properly next time). But I think this might be tricky given all the possible paths, so it's probably not worth it.

Reply:
That would be tricky to do and test - I don't think it's worth the effort tbh. Already we have twice as much code for making error messages as actually doing the work.